deel/lip/metrics.py

```diff
@@ -0,0 +1,204 @@
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+from tensorflow.keras.losses import Reduction
+from tensorflow.keras.utils import register_keras_serializable
+
+
+def _delta_multiclass(y_true, y_pred):
+    r"""
+    Compute the non-normalized provable robustness factor in multiclass setup (need
+    to be adjusted with
+    lipschitz constant.
+
+    .. math::
+            \Delta(x) = f_l(x) - \max_{i \neq l} f_i(x)
+
+    Args:
+        y_true: true labels must be in {1,0} or in {1,-1} (no label smoothing allowed)
+        y_pred: network predictions
+
+    Returns: non-normalized provable robustness factor
+
+    """
+    ynl_shape = (-1, tf.shape(y_pred)[-1] - 1)
+    yl = tf.boolean_mask(y_pred, y_true == 1)
+    ynl = tf.reshape(
+        tf.boolean_mask(y_pred, y_true != 1),
+        ynl_shape,
+    )
+    delta = yl - tf.reduce_max(ynl, axis=-1)
+    return delta
+
+
+def _delta_binary(y_true, y_pred):
+    r"""
+    Compute the non-normalized provable robustness factor in binary setup (need to be
+    adjusted with
+    lipschitz constant).
+
+    .. math::
+            \Delta(x) = f(x) \text{ if } l=1, -f(x) \text{ otherwise}
+
+    Args:
+        y_true: true labels must be in {1,0} or in {1,-1} (no label smoothing allowed)
+        y_pred: network predictions
+
+    Returns: non-normalized provable robustness factor
+
+    """
+    y_true = tf.sign(tf.cast(y_true, y_pred.dtype) - 1e-3)
+    return tf.multiply(y_true, y_pred)
+
+
+@register_keras_serializable("deel-lip", "ProvableRobustAccuracy")
+class ProvableRobustAccuracy(Loss):
+    def __init__(
+        self,
+        epsilon=36 / 255,
+        lip_const=1.0,
+        disjoint_neurons=True,
+        reduction=Reduction.AUTO,
+        name="ProvableRobustAccuracy",
+    ):
+        r"""
+
+        The accuracy that can be proved at a given epsilon.
+
+        Args:
+            epsilon: the metric will return the guaranteed accuracy for the radius
+                epsilon
+            lip_const: lipschitz constant of the network
+            disjoint_neurons: must be set to True is your model ends with a
+                FrobeniusDense layer with `disjoint_neurons` set to True. Set to False
+                otherwise
+            reduction: the recution method when training in a multi-gpu / TPU system
+            name: metrics name.
+        """
+        self.lip_const = lip_const
+        self.epsilon = epsilon
+        self.disjoint_neurons = disjoint_neurons
+        if disjoint_neurons:
+            self.certificate_factor = 2 * lip_const
+        else:
+            self.certificate_factor = math.sqrt(2) * lip_const
+        super(ProvableRobustAccuracy, self).__init__(reduction, name)
+
+    def call(self, y_true, y_pred):
+        shape = y_true.shape
+        if len(shape) == 2 and (shape[-1] > 1):
+            delta_fct = _delta_multiclass
+        else:
+            delta_fct = _delta_binary
+            self.certificate_factor = self.lip_const
```

Comment:
That's sound odd to me to set the `self.certificate_factor` here in the `call()`. I don't know what is the simplest and cleanest option... Maybe an argument for binary vs. multiclass setup? Or two separate metrics, such as in losses.py?

Reply:
yeah, in combination with a `@tf.function`, `y_true.shape` is considered as a constant and thus the built graph don't embed the `if`, and there is no overhead. This wouldn't be the case if we used a `tf.shape(y_true)` (which is an operator).